React/Testing best practices.md

```diff
@@ -0,0 +1,610 @@
+
+```tsx
+const AllProviders = ({ children }) => (
+  <ChakraProvider theme={theme}>{children}</ChakraProvider>
+);
+```
+
+Beside providers we need to mock [datx](https://datx.dev/) store
+
+```tsx
+export const StoreMockContext = React.createContext<AppCollection | null>(null);
+
+const withMockStore = (PageComponent: NextPage) => {
+  const WithMockStore: FC = (props) => {
+    const store: AppCollection = new AppCollection();
+
+    return (
+      <StoreMockContext.Provider value={store}>
+        <PageComponent {...props} />
+      </StoreMockContext.Provider>
+    );
+  };
+
+  return WithMockStore;
+};
+```
+
+Now while testing we can wrap a component in `<StoreMockContext.Consumer>` to get access to the store. Store is needed when we need to mock datx model with relationships.
+
+After mocking all providers and store all needs to be exported.
+
+```tsx
+const customRender = (ui: React.ReactElement, options?: any) =>
+  render(ui, { wrapper: withMockStore(AllProviders), ...options });
+
+// re-export everything
+export * from "@testing-library/react";
+
+// override render method
+export { customRender as render };
+```
+
+Now when testing components we won't import `render` from `@testing-library/react` but instead from `__tests__/test-utils.tsx` like this:
+
+```tsx
+import { render } from "__tests__/test-utils";
+```
+
+Or if we add `path` to `tsconfig`:
+
+```json
+{
+  "compilerOptions": {
+    ...
+    "paths": {
+      ...
+      "@test-utils": ["__tests__/test-utils.tsx"]
+    }
+  },
+}
+```
+
+Then we can import render like this:
+
+```tsx
+import { render } from "@test-utils";
+```
+
+### Folder structure
+
+All tests should be defined in the same directory where the file being tested is. There is one exception and that is `pages` folder because Next.js doesn't allow tests in `pages` folder. Thats why we have `__tests__` folder in the root of our application.
+
+Example:
+
+```
+src
+.
+├── __mocks__
+│   └── react-i18next.tsx
+├── __tests__
+│   ├── pages
+│   │   └── user.test.ts
+│   └── test-utils.tsx
+├── pages
+│   └── user.ts
+├── fetchers
+│   └── users
+│      ├── users.ts
+│      └── users.test.ts
+└── components
+    └── ui
+        ├── atoms
+        │   └── button
+        │       ├── Button.test.tsx
+        │       └── Button.tsx
+        └── molecules
+            └── user-card
+                ├── UserCard.test.tsx
+                └── UserCard.tsx
+```
+
+### `__mocks__`
+
+Manual mocks are used to stub out functionality with mock data. For example, instead of accessing a resource from `node_modules` you might want to create a mock module that allows you to use fake data. Manual mocks are defined by writing a module in a `__mocks__/` subdirectory immediately adjacent to the module. ([docs](https://jestjs.io/docs/en/manual-mocks))
```

Comment:
```suggestion
Manual mocks are used to stub out functionality with mock data. For example, instead of accessing a resource from `node_modules` you might want to create a mock module that allows you to use fake data. Manual mocks are defined by writing a module in the `__mocks__/` subdirectory immediately adjacent to the module. ([docs](https://jestjs.io/docs/en/manual-mocks))
```